src/cpp/statistics/rtps/StatisticsBase.hpp

```diff
@@ -0,0 +1,287 @@
+        }
+
+    };
+
+    using GUID_t = fastrtps::rtps::GUID_t;
+    using Key = std::shared_ptr<ListenerProxy>;
+
+    // specialized comparison operator, the actual key is the external listener address
+    struct CompareProxies
+    {
+        bool operator ()(
+                const Key& left,
+                const Key& right) const
+        {
+            return *left < *right;
+        }
+
+    };
+
+    using ProxyCollection = std::set<Key, CompareProxies>;
+    ProxyCollection listeners_;
+
+    // retrieve the participant mutex
+    std::recursive_mutex& get_statistics_mutex();
+
+    /** Register a listener in participant RTPSWriter entities.
+     * @param listener, smart pointer to the listener interface to register
+     * @param guid, RTPSWriter identifier. If unknown the listener is registered in all enable ones
+     * @return true on success
+     */
+    virtual bool register_in_writer(
+            std::shared_ptr<fastdds::statistics::IListener> listener,
+            GUID_t guid = GUID_t::unknown()) = 0;
+
+    /** Register a listener in participant RTPSReader entities.
+     * @param listener, smart pointer to the listener interface to register
+     * @param guid, RTPSReader identifier. If unknown the listener is registered in all enable ones
+     * @return true on success
+     */
+    virtual bool register_in_reader(
+            std::shared_ptr<fastdds::statistics::IListener> listener,
+            GUID_t guid = GUID_t::unknown()) = 0;
+
+    /** Unregister a listener in participant RTPSWriter entities.
+     * @param listener, smart pointer to the listener interface to unregister
+     * @return true on success
+     */
+    virtual bool unregister_in_writer(
+            std::shared_ptr<fastdds::statistics::IListener> listener) = 0;
+
+    /** Unregister a listener in participant RTPSReader entities.
+     * @param listener, smart pointer to the listener interface to unregister
+     * @return true on success
+     */
+    virtual bool unregister_in_reader(
+            std::shared_ptr<fastdds::statistics::IListener> listener) = 0;
+
+    // lambda function to traverse the listener collection
+    template<class Function>
+    Function for_each_listener(
+            Function f)
+    {
+        std::lock_guard<std::recursive_mutex> lock(get_statistics_mutex());
+
+        for (auto& listener : listeners_)
+        {
+            f(listener);
+        }
+
+        return f;
+    }
+
+    // returns if a mask statistics::EventKind may require participant writers update
+    bool are_writers_involved(
+            const uint32_t mask) const;
+
+    // returns if a mask statistics::EventKind may require participant readers update
+    bool are_readers_involved(
+            const uint32_t mask) const;
+
+    // TODO: methods for listeners callbacks
+
+    /*
+     * Report that a message that is sent by the participant
```

Comment:
```suggestion
     * Report a message that is sent by the participant
```

OR

```suggestion
     * Report that a message is sent by the participant
```

Reply:
What I was thinking? I pick the second one